urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+    // e.g: <del>abcxxx</del><ins>xxxdef</ins>
+    //   -> <del>abc</del>xxx<ins>def</ins>
+    // e.g: <del>xxxabc</del><ins>defxxx</ins>
+    //   -> <ins>def</ins>xxx<del>abc</del>
+    // Only extract an overlap if it is as big as the edit ahead or behind it.
+    pointer = diffs.listIterator();
+    Diff prevDiff = null;
+    thisDiff = null;
+    if (pointer.hasNext()) {
+      prevDiff = pointer.next();
+      if (pointer.hasNext()) {
+        thisDiff = pointer.next();
+      }
+    }
+    while (thisDiff != null) {
+      if (prevDiff.operation == Operation.DELETE &&
+          thisDiff.operation == Operation.INSERT) {
+        String deletion = prevDiff.text;
+        String insertion = thisDiff.text;
+        int overlap_length1 = this.diff_commonOverlap(deletion, insertion);
+        int overlap_length2 = this.diff_commonOverlap(insertion, deletion);
+        if (overlap_length1 >= overlap_length2) {
+          if (overlap_length1 >= deletion.length() / 2.0 ||
+              overlap_length1 >= insertion.length() / 2.0) {
+            // Overlap found. Insert an equality and trim the surrounding edits.
+            pointer.previous();
+            pointer.add(new Diff(Operation.EQUAL,
+                                 insertion.substring(0, overlap_length1)));
+            prevDiff.text =
+                deletion.substring(0, deletion.length() - overlap_length1);
+            thisDiff.text = insertion.substring(overlap_length1);
+            // pointer.add inserts the element before the cursor, so there is
+            // no need to step past the new element.
+          }
+        } else {
+          if (overlap_length2 >= deletion.length() / 2.0 ||
+              overlap_length2 >= insertion.length() / 2.0) {
+            // Reverse overlap found.
+            // Insert an equality and swap and trim the surrounding edits.
+            pointer.previous();
+            pointer.add(new Diff(Operation.EQUAL,
+                                 deletion.substring(0, overlap_length2)));
+            prevDiff.operation = Operation.INSERT;
+            prevDiff.text =
+              insertion.substring(0, insertion.length() - overlap_length2);
+            thisDiff.operation = Operation.DELETE;
+            thisDiff.text = deletion.substring(overlap_length2);
+            // pointer.add inserts the element before the cursor, so there is
+            // no need to step past the new element.
+          }
+        }
+        thisDiff = pointer.hasNext() ? pointer.next() : null;
+      }
+      prevDiff = thisDiff;
+      thisDiff = pointer.hasNext() ? pointer.next() : null;
+    }
+  }
+
+  /**
+   * Look for single edits surrounded on both sides by equalities
+   * which can be shifted sideways to align the edit to a word boundary.
+   * e.g: The c<ins>at c</ins>ame. -> The <ins>cat </ins>came.
+   * @param diffs LinkedList of Diff objects.
+   */
+  public void diff_cleanupSemanticLossless(LinkedList<Diff> diffs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.